palace/drivers/basesolver.cpp

```diff
@@ -84,6 +89,282 @@ BaseSolver::BaseSolver(const IoData &iodata, bool root, int size, int num_thread
+      }
+    }
+    else if (param.write_serial_mesh)
+    {
+      if (mesh->Nonconforming())
+      {
+        // Given no rebalancing will be done, need to handle the serial write more
+        // carefully. For NC, this requires creating a duplicate mesh.
+        mfem::ParMesh smesh(*mesh);
+        mfem::Array<int> serial_partition(mesh->GetNE());
+        serial_partition = 0;
+        smesh.Rebalance(serial_partition);
+        BlockTimer bt_io(Timer::IO);
+        if (Mpi::Root(comm))
+        {
+          std::ofstream serial(serial_mesh_filename);
+          serial.precision(mesh::MSH_FLT_PRECISION);
+          mesh::DimensionalizeMesh(smesh, iodata.GetMeshScaleFactor());
+          smesh.Mesh::Print(serial);  // Do not need to nondimensionalize the temporary
+        }
+        Mpi::Barrier(comm);
+      }
+      else
+      {
+        auto smesh = std::make_unique<mfem::Mesh>(mesh->GetSerialMesh(0));
+        BlockTimer bt(Timer::IO);
+        if (Mpi::Rank(comm) == 0)
+        {
+          std::ofstream serial(serial_mesh_filename);
+          serial.precision(mesh::MSH_FLT_PRECISION);
+          mesh::DimensionalizeMesh(*smesh, iodata.GetMeshScaleFactor());
+          smesh->Print(serial);  // Do not need to nondimensionalize the temporary
+        }
+      }
+    }
+    mesh->ExchangeFaceNbrData();
+  }
+}
+
+}  // namespace
+
+void BaseSolver::SolveEstimateMarkRefine(
+    std::vector<std::unique_ptr<mfem::ParMesh>> &mesh) const
+{
+  const auto &param = iodata.model.refinement;
+  auto comm = mesh.back()->GetComm();
+  // Helper to save off postprocess data.
+  auto SavePostProcess = [&](int iter)
+  {
+    Mpi::Barrier(comm);
+    BlockTimer bt_io(Timer::IO);
+    if (Mpi::Root(comm) && param.save_adapt_iterations)
+    {
+      // Create a subfolder for the results of this adaptation.
+      namespace fs = std::filesystem;
+      const auto out_dir = fs::path(IterationPostDir(iter));
+      constexpr auto options =
+          fs::copy_options::recursive | fs::copy_options::overwrite_existing;
+      const fs::path root_dir(post_dir);
+      for (const auto &f : fs::directory_iterator(root_dir))
+      {
+        if (f.is_regular_file())
+        {
+          fs::copy(f, out_dir / fs::path(f).filename(), options);
+        }
+      }
+      if (fs::exists(root_dir / "paraview"))
+      {
+        fs::copy(root_dir / "paraview", out_dir / "paraview", options);
+      }
+    }
+    Mpi::Barrier(comm);
+  };
+
+  const bool use_amr = param.adapt_max_its > 0;
+  const bool use_coarsening = mesh.back()->Nonconforming() && param.use_coarsening;
+  if (use_amr && mesh.size() > 1)
+  {
+    Mpi::Print("{}\n", "Flattening mesh sequence: AMR will start from the final mesh in "
+                       "the refinement sequence.");
+    mesh.erase(mesh.begin(), mesh.end() - 1);
+    constexpr bool refine = true, fix_orientation = true;
+    mesh.back()->Finalize(refine, fix_orientation);
```

Comment:
Can you not assume this has already been called during the load in `geodata.cpp`?

Reply:
The issue is that `UniformRefinement()` of a tet mesh invalidates the finalization. From `mesh.hpp`:
```
   /** @param[in] ref_algo %Refinement algorithm. Currently used only for pure
       tetrahedral meshes. If set to zero (default), a tet mesh will be refined
       using algorithm A, that produces elements with better quality compared to
       algorithm B used when the parameter is non-zero.

       For tetrahedral meshes, after using algorithm A, the mesh cannot be
       refined locally using methods like GeneralRefinement() unless it is
       re-finalized using Finalize() with the parameter @a refine set to true.
       Note that calling Finalize() in this way will generally invalidate any
       FiniteElementSpace%s and GridFunction%s defined on the mesh. */
   void UniformRefinement(int ref_algo = 0);
```
The issue being I believe that the uniform refinement uses another code path that doesn't respect the edge ordering . Hence the finalization only being on the pathway where the vector contains meshes generated from refinement.

Reply:
Ah, makes sense to me. Thanks!